Refactors the Sliproad Handler

daniel-j-h opened the pull request:
Refactors the Sliproad Handler.

cc @MoKob

MoKob commented on src/extractor/guidance/sliproad_handler.cpp:
```diff
@@ -34,221 +36,152 @@ SliproadHandler::SliproadHandler(const IntersectionGenerator &intersection_gener
+    {
+        auto same_road_category =
+            next_road_edge_data.road_classification == source_edge_data.road_classification;
+        auto same_name = next_road_edge_data.name_id != EMPTY_NAMEID &&
```
should we use `requiresNameAnnounced` here as well?

daniel-j-h replied:
Yeah this will be done in https://github.com/Project-OSRM/osrm-backend/pull/2945 - I just haven't rebased and finished it yet.

There will be some conflicts, for sure.

daniel-j-h commented:
Squashed and rebased against master - starting to sample real-world situations now looking for both false positives and negatives.

daniel-j-h commented:
Here's what we currently classify as sliproad scenarios in Berlin:

https://api.mapbox.com/styles/v1/danieljh/ciuyc3mrp01c92jl852h1ntu5.html?title=true&access_token=pk.eyJ1IjoiZGFuaWVsamgiLCJhIjoiTnNYb25JSSJ9.vYOnsuu1zeKcGW2nj0uJZw#12.17/52.5081/13.3670

MoKob commented:
Glad we have this visualisation, there quite a few strange ones in there

daniel-j-h commented:
Now with restricting the Sliproad classification to ramps and links:

https://api.mapbox.com/styles/v1/danieljh/ciuznn1mg01ga2iqqpdi9rsr4.html?title=true&access_token=pk.eyJ1IjoiZGFuaWVsamgiLCJhIjoiTnNYb25JSSJ9.vYOnsuu1zeKcGW2nj0uJZw#12.17/52.5081/13.3670

daniel-j-h commented:
Latest visualization is here:

https://api.mapbox.com/styles/v1/danieljh/civ2i4gf000402ino5qir8u1r.html?title=true&access_token=pk.eyJ1IjoiZGFuaWVsamgiLCJhIjoiTnNYb25JSSJ9.vYOnsuu1zeKcGW2nj0uJZw#11.07/52.5107/13.4348

Pink are old(er) commits, blue is current state:
- taking link / ramp into account
- fixes for discarding Forks
- narrow angle for turn off the Sliproad and back onto the street

Next checks to try out:
- [ ] turn off of the Sliproad `ade` has to be obvious (implies entry_allowed)
- [ ] non-sliproad `cde` has to be straight-ish (~15 degree) for turn at intersection `c` where sliproad joins
- [ ] test curvature, sliproad `ad` has to snuggle against roads it shortcuts `abc` `cde` (check ccw / cw, fix start `a`, end `d`)

General fixes that need to land:
- [ ] Graph walker for skipping traffic lights on all roads involved
- [ ] Graph walker for finding next intersection / distance check

---------------------------------------------------------------------------------------------------------

Some naming conventions used here

```
       .
       e
       .
       .
 a ... b .... c .   < `next` is intersection at `c`
       `      .
         `    .
           `  .
              d     < `target_intersection` is intersection at `d`
              .
              e


          ^ `sliproad` is `bd`
       ^ `intersection` is intersection at `b`
```

daniel-j-h commented on src/extractor/guidance/sliproad_handler.cpp:
```diff
@@ -135,13 +135,13 @@ operator()(const NodeID /*nid*/, const EdgeID source_edge_id, Intersection inter
 
-    std::unordered_set<NameID> target_road_names;
+    std::vector<NameID> target_road_name_ids;
+    target_road_name_ids.reserve(next->intersection.size());
```
@MoKob I checked the sizes on Berlin, turns out it's really small.
We're fine with reserving here and then simply looping below.

Ticketed the `string_ref` idea in https://github.com/Project-OSRM/osrm-backend/issues/3265.

MoKob replied:
Yeah. They are small. I just remember someone telling me to used `unordered_sets` for these ;)

daniel-j-h commented:
Latest snapshot:

- [Berlin](https://api.mapbox.com/styles/v1/danieljh/civ2i4gf000402ino5qir8u1r.html?title=true&access_token=pk.eyJ1IjoiZGFuaWVsamgiLCJhIjoiTnNYb25JSSJ9.vYOnsuu1zeKcGW2nj0uJZw#10.85/52.5121/13.4983)
- [Bay Area](https://api.mapbox.com/styles/v1/danieljh/civ2i4gf000402ino5qir8u1r.html?title=true&access_token=pk.eyJ1IjoiZGFuaWVsamgiLCJhIjoiTnNYb25JSSJ9.vYOnsuu1zeKcGW2nj0uJZw#10.29/37.7060/-122.2671)

There are some ideas left to explore as noted in https://github.com/Project-OSRM/osrm-backend/pull/3199#issuecomment-258197682.

daniel-j-h commented:
Here is a current snapshot with additional heuristics implemented:

https://api.mapbox.com/styles/v1/danieljh/civ2i4gf000402ino5qir8u1r.html?title=true&access_token=pk.eyJ1IjoiZGFuaWVsamgiLCJhIjoiTnNYb25JSSJ9.vYOnsuu1zeKcGW2nj0uJZw#10.411588333960331/52.51237533716511/13.416338076773059/0

@MoKob - Red before, Green at the moment.
Also check out SF for how it looks currently.

MoKob commented:
This looks way better, I've found a few issues though:

- [ ] should be possible to filter out given the size of the intersection at sliproad end (here 4). I'd expect a sliproad to have intersection size 3. Did you check for this? Or did you find it to be different?
https://api.mapbox.com/styles/v1/danieljh/civ2i4gf000402ino5qir8u1r.html?title=true&access_token=pk.eyJ1IjoiZGFuaWVsamgiLCJhIjoiTnNYb25JSSJ9.vYOnsuu1zeKcGW2nj0uJZw#17.47/52.51539/13.59450

- [ ] we could possibly limit the minimum street category, or at least make low street-categories only consider very low distances
https://api.mapbox.com/styles/v1/danieljh/civ2i4gf000402ino5qir8u1r.html?title=true&access_token=pk.eyJ1IjoiZGFuaWVsamgiLCJhIjoiTnNYb25JSSJ9.vYOnsuu1zeKcGW2nj0uJZw#17.56/52.61139/13.48635
https://api.mapbox.com/styles/v1/danieljh/civ2i4gf000402ino5qir8u1r.html?title=true&access_token=pk.eyJ1IjoiZGFuaWVsamgiLCJhIjoiTnNYb25JSSJ9.vYOnsuu1zeKcGW2nj0uJZw#17.85/52.56741/13.21783

- [ ] the edge is the sliproad, right here the road does not allow the proposed turn, even without the turn restriction. This could be easily checked by looking at the `reversed` flag. A road that connects to the sliproad turn cannot be reversed. https://api.mapbox.com/styles/v1/danieljh/civ2i4gf000402ino5qir8u1r.html?title=true&access_token=pk.eyJ1IjoiZGFuaWVsamgiLCJhIjoiTnNYb25JSSJ9.vYOnsuu1zeKcGW2nj0uJZw#19.16/52.42785/13.40512 looks like the turn to the `left` does only allow a u-turn here. 

- [ ] the sliproad should not be the end of a different edge:
https://api.mapbox.com/styles/v1/danieljh/civ2i4gf000402ino5qir8u1r.html?title=true&access_token=pk.eyJ1IjoiZGFuaWVsamgiLCJhIjoiTnNYb25JSSJ9.vYOnsuu1zeKcGW2nj0uJZw#17.66/52.37530/13.39308

MoKob commented on features/guidance/advanced-lanes.feature:
```diff
@@ -10,8 +10,11 @@ Feature: Turn Lane Guidance
-                          d
+                          .
+            a ... b ..... c . g
+                    `     .
```
modification of test - intended? this increases distance between `c` and `d`

MoKob commented on features/guidance/advanced-lanes.feature:
```diff
@@ -41,8 +44,10 @@ Feature: Turn Lane Guidance
-                          d
+            a . . b . . . c g
+                    `     .
+                       `  .
```
even more so here. Please check all test modifications whether they are intended or not

MoKob commented on features/guidance/collapse.feature:
```diff
@@ -725,10 +725,10 @@ Feature: Collapse
-            | a,g       | road,cross,cross      | depart,turn left,arrive                    |
-            | a,e       | road,road,road        | depart,continue straight,arrive            |
+            | a,g       | road,cross,cross      | depart,fork left,arrive                    |
+            | a,e       | road,road,road        | depart,fork slight right,arrive            |
```
this test here was supposed to collapse the turns `bcg` and `bde`,`bdf` which now does not happen anymore.

This change modifies the purpose of the test.

MoKob replied:
The scenario in question still provides a now different test. Can you create a copy of this test with similar setup that again identifies a sliproad here? The feature in collapse is supposed to collapse, which is not happening now.
The test as it is here can be moved to a different feature set.

MoKob commented on include/engine/api/route_api.hpp:
```diff
@@ -152,6 +153,8 @@ class RouteAPI : public BaseAPI
 
                 guidance::trimShortSegments(steps, leg_geometry);
+                // TODO: remove before merging
+                util::guidance::print(steps);
```
don't forget :)

MoKob commented on include/extractor/guidance/sliproad_handler.hpp:
```diff
@@ -9,10 +9,15 @@
 #include "util/node_based_graph.hpp"
 
+#include "util/geojson_debug_logger.hpp"
+#include "util/geojson_debug_policies.hpp"
```
remove before merge

MoKob commented on include/extractor/guidance/sliproad_handler.hpp:
```diff
@@ -42,6 +47,50 @@ class SliproadHandler : public IntersectionHandler
+
+  private:
+    util::ScopedGeojsonLoggerGuard<util::NodeIdVectorToLineString> geojson_lines;
+    util::ScopedGeojsonLoggerGuard<util::NodeIdVectorToMultiPoint> geojson_points;
```
remove before merge

MoKob commented on include/extractor/guidance/sliproad_handler.hpp:
```diff
@@ -42,6 +47,50 @@ class SliproadHandler : public IntersectionHandler
+    };
+
+    // Returns a potential next intersection along a road skipping over traffic lights.
+    boost::optional<IntersectionAndNode> getNextIntersection(const NodeID at,
```
should probably not be a part of the sliproad handler but at least of the base class, allowing all handlers to find next intersections

MoKob commented on src/extractor/guidance/intersection.cpp:
```diff
@@ -112,6 +117,45 @@ Intersection::getHighestConnectedLaneCount(const util::NodeBasedDynamicGraph &gr
     boost::range::find_if(view, extract_maximal_value);
     return max_lanes;
+
+    return max_lanes;
```
duplicated return?

MoKob commented on src/extractor/guidance/intersection.cpp:
```diff
@@ -112,6 +117,45 @@ Intersection::getHighestConnectedLaneCount(const util::NodeBasedDynamicGraph &gr
+
+ConnectedRoad &Intersection::getUTurnRoad()
+{
+    BOOST_ASSERT(valid());
```
while technically reasonable, checking the validity at every such iterator could have a serious performance impact on debug builds/release with assertion builds.

I would say that we should rely on the generators to assert validity to not run into problems with debug builds/assertion builds on the demo server.

MoKob commented on src/extractor/guidance/sliproad_handler.cpp:
```diff
@@ -29,273 +32,720 @@ SliproadHandler::SliproadHandler(const IntersectionGenerator &intersection_gener
-                          intersection_generator)
+                          intersection_generator),
+      geojson_lines{"sliproads-lines.geojson", node_info_list},
+      geojson_points{"sliproads-points.geojson", node_info_list}
```
remove before merge

MoKob commented on src/extractor/guidance/sliproad_handler.cpp:
```diff
@@ -29,273 +32,720 @@ SliproadHandler::SliproadHandler(const IntersectionGenerator &intersection_gener
-    if (intersection.size() <= 2)
-        return intersection;
+    // Road index prefering non-sliproads (bc)
+    auto obvious = getObviousIndexWithSliproads(source_edge_id, intersection, intersection_node_id);
```
since this returns an optional, I feel that `getObviousIndex` could be misleading as a name. We do not have a naming convention here, so far as I know. But how about `checkIfHasObviousIndexWithSliproads`?

MoKob commented on src/extractor/guidance/sliproad_handler.cpp:
```diff
@@ -29,273 +32,720 @@ SliproadHandler::SliproadHandler(const IntersectionGenerator &intersection_gener
-                return true;
-            };
+    // Potential non-sliproad road (bc), leading to the intersection (c) the Sliproad (bd) shortcuts
+    const auto &next_road = intersection[*obvious];
```
could probably also benefit from an improvement in naming over my original poor choice

MoKob commented on src/extractor/guidance/sliproad_handler.cpp:
```diff
@@ -29,273 +32,720 @@ SliproadHandler::SliproadHandler(const IntersectionGenerator &intersection_gener
-    }();
+    // The intersection `c` has to be reachable
+    if (!next_road.entry_allowed)
+    {
```
Are we using `{}` for single line `if` branches, or not? I feel that this is often inconsistent in our code-base. Just 5 lines above you use the other style. I'd prefer not having the additional braces here.

MoKob commented on src/extractor/guidance/sliproad_handler.cpp:
```diff
@@ -29,273 +32,720 @@ SliproadHandler::SliproadHandler(const IntersectionGenerator &intersection_gener
-    if (obvious_turn_index == 0)
+    // The road leading to the intersection (bc) has to continue from our source
+    if (!roadContinues(source_edge_id, next_road.eid))
+    {
```
and here

MoKob commented on src/extractor/guidance/sliproad_handler.cpp:
```diff
@@ -29,273 +32,720 @@ SliproadHandler::SliproadHandler(const IntersectionGenerator &intersection_gener
-        std::find_if(intersection.begin(), intersection.end(), linkTest) != intersection.end();
-    if (!hasNarrow)
-        return intersection;
+        if (!is_narrow || same_angle)
```
you've probably spent some time here, but what about an intersection that models a turn when the road turns?
```
           .   d
a - b
           `   c
```

Why would this not be a sliproad when

```
            . d
        .
a - b - - - - c
```

could be?

MoKob replied:
I've seen quite a selection of intersections that turn right before merging into a road. So I'd say we should try and detect these.
I'd say that the main road has to be narrow and the sliproad can't turn far from it.

MoKob commented on src/extractor/guidance/sliproad_handler.cpp:
```diff
@@ -29,273 +32,720 @@ SliproadHandler::SliproadHandler(const IntersectionGenerator &intersection_gener
+            return false;
+
+        // Prevent from starting in or going onto a roundabout
+        auto onto_roundabout = hasRoundaboutType(road.instruction);
```
this check should probably be done earlier, since it is a simple check on the instruction and does not require looking for an obvious index (which costs more)

MoKob commented on src/extractor/guidance/sliproad_handler.cpp:
```diff
@@ -29,273 +32,720 @@ SliproadHandler::SliproadHandler(const IntersectionGenerator &intersection_gener
-        return road_edge_data.road_classification == source_edge_data.road_classification &&
-               road_edge_data.name_id != EMPTY_NAMEID && source_edge_data.name_id != EMPTY_NAMEID &&
-               same_name && road.entry_allowed;
+        auto is_roundabout = road_data.roundabout;
```
duplicated check? Within roundabouts, you should have only size two connections.
Outside of roundabouts/on exits you would see a roundabout type instruction.

Also, If it should be necessary, I'd prefer doing a single function style check for `isRoundabout` instead of two separate checks.

MoKob commented on src/extractor/guidance/sliproad_handler.cpp:
```diff
@@ -29,273 +32,720 @@ SliproadHandler::SliproadHandler(const IntersectionGenerator &intersection_gener
-    if (next_road_length > MAX_SLIPROAD_THRESHOLD)
+    // If the intersection is too far away, don't bother continuing
+    if (nextIntersectionIsTooFarAway(intersection_node_id, next_road.eid))
     {
```
brace style

MoKob commented on src/extractor/guidance/sliproad_handler.cpp:
```diff
@@ -29,273 +32,720 @@ SliproadHandler::SliproadHandler(const IntersectionGenerator &intersection_gener
-        &util::coordinate_calculation::haversineDistance);
-    if (next_road_length > MAX_SLIPROAD_THRESHOLD)
+    // If the intersection is too far away, don't bother continuing
+    if (nextIntersectionIsTooFarAway(intersection_node_id, next_road.eid))
```
is this check valid? Looks like you are finding the next intersection below. You need to know about the full distance for this check, don't you? Or are you calculating the path to the next intersection twice? (In that case, it could be considered performance bug)

MoKob commented on src/extractor/guidance/sliproad_handler.cpp:
```diff
@@ -29,273 +32,720 @@ SliproadHandler::SliproadHandler(const IntersectionGenerator &intersection_gener
+    target_road_name_ids.reserve(next->intersection.size());
 
-    for (const auto &road : next_road_next_intersection)
+    for (const auto &road : next->intersection)
```
can we potentially change `next` into `path_along_main_road` or something similar?

MoKob commented on src/extractor/guidance/sliproad_handler.cpp:
```diff
@@ -29,273 +32,720 @@ SliproadHandler::SliproadHandler(const IntersectionGenerator &intersection_gener
 
-    for (auto &road : intersection)
+    // TODO: remove geojson debugging
+    NodeID b, c, d;
```
Please never use `b`,`c`,d` or other single-char variable names. Finding where they are used if you try to follow the code is near impossible. If you need these to make your cases simple, I'd prefer adding a lookup table at the top of the examples. (e.g. node_at_intersection (c), ...)

MoKob commented on src/extractor/guidance/sliproad_handler.cpp:
```diff
@@ -29,273 +32,720 @@ SliproadHandler::SliproadHandler(const IntersectionGenerator &intersection_gener
+    for (std::size_t road_index = 0, last = intersection.size(); road_index < last; ++road_index)
     {
-        if (linkTest(road))
+        auto &sliproad = intersection[road_index]; // this is what we're checking and assigning to
```
`possible_sliproad` or `sliproad_candidate`?

MoKob commented on src/extractor/guidance/sliproad_handler.cpp:
```diff
@@ -29,273 +32,720 @@ SliproadHandler::SliproadHandler(const IntersectionGenerator &intersection_gener
+            continue;
+
+        // b -> d edge id
+        EdgeID sliproad_edge = sliproad.eid;
```
`auto`?

MoKob commented on src/extractor/guidance/sliproad_handler.cpp:
```diff
@@ -29,273 +32,720 @@ SliproadHandler::SliproadHandler(const IntersectionGenerator &intersection_gener
+        EdgeID sliproad_edge = sliproad.eid;
+
+        // TODO: remove debugging
+        d = node_based_graph.GetTarget(sliproad_edge);
```
should be something like `node_at_end_of_sliproad`

MoKob commented on src/extractor/guidance/sliproad_handler.cpp:
```diff
@@ -29,273 +32,720 @@ SliproadHandler::SliproadHandler(const IntersectionGenerator &intersection_gener
+        d = node_based_graph.GetTarget(sliproad_edge);
+
+        const auto target_intersection = [&](NodeID node) {
+            auto intersection = intersection_generator(node, sliproad_edge);
```
don't you have logic for skipping over these? Also: there is a high potential for multiple traffic lights. The intersection_generator as well offers `GetActualNextIntersection` (could use a better name) which does exactly what you do here, a bit more efficiently. Here you are generating coordinates for all traffic lights, but you don't care about these.

MoKob commented on src/extractor/guidance/sliproad_handler.cpp:
```diff
@@ -29,273 +32,720 @@ SliproadHandler::SliproadHandler(const IntersectionGenerator &intersection_gener
+            return intersection;
+        }(intersection_node_id);
+
+        const NodeID sliproad_edge_target = node_based_graph.GetTarget(sliproad_edge);
```
you now have `d` and `sliproad_edge_target` both from `sliproad_edge`. Following these without better names can become difficult. I'd prefer `first_node_along_sliproad_candidate` and `node_at_next_intersection_along_sliproad`

MoKob commented on src/extractor/guidance/sliproad_handler.cpp:
```diff
@@ -29,273 +32,720 @@ SliproadHandler::SliproadHandler(const IntersectionGenerator &intersection_gener
+
+        // Distinct triangle nodes `bcd`
+        if (intersection_node_id == next->node || intersection_node_id == sliproad_edge_target ||
+            next->node == sliproad_edge_target)
```
how does `next->node` interact with `d` and `sliproad_edge_target`?
Wouldn't `intersection_node_id == sliproad_edge_target` imply a self-loop? 😕

MoKob commented on src/extractor/guidance/sliproad_handler.cpp:
```diff
@@ -29,273 +32,720 @@ SliproadHandler::SliproadHandler(const IntersectionGenerator &intersection_gener
+
+        // The turn off of the Sliproad has to be obvious and a narrow turn
+        {
+            const auto index = findObviousTurn(sliproad_edge, target_intersection);
```
I'd go as far as to say it has to be the `only` turn available. Did you find any coincidences where larger intersections would be desirable?

MoKob commented on src/extractor/guidance/sliproad_handler.cpp:
```diff
@@ -29,273 +32,720 @@ SliproadHandler::SliproadHandler(const IntersectionGenerator &intersection_gener
+        //
+        //                    Sliproad           Not a Sliproad
+        {
+            // Intersection is orderd: 0 is UTurn, then from sharp right to sharp left.
```
some of these checks are quite expensive. Have you checked whether you can re-order some of them so that cheaper ones come first to exit early when possible?

Can you potentially move some of these into dedicated functions (detectors) and annotate them in `O(n)`. That could make finding a reasonable order easier.

MoKob commented on src/extractor/guidance/sliproad_handler.cpp:
```diff
@@ -29,273 +32,720 @@ SliproadHandler::SliproadHandler(const IntersectionGenerator &intersection_gener
+        //              d
+        //
+        const auto area_threshold = std::pow(
+            scaledThresholdByRoadClass(MAX_SLIPROAD_THRESHOLD, sliproad_data.road_classification),
```
assuming the length of the sliproad scales linearly, I'd expect `scaledThreshold()/2` here to have a more consistent relation between sliproadLength (which is the threshold, right?) and triangle are.

MoKob replied:
Does this really change so much? It's a bit confusing to be, but ok to keep. It's just that we are looking at a triangle area but filter by the square. So a longer sliproad allows stronger growing areas, which would result in scenarios that scale up linearly to work for one length and fail for the other.

e.g.
```
a - b - c
        d
```
could fail for a smaller grid size and work for a larger one.
Or is my math off here?

MoKob commented on src/extractor/guidance/sliproad_handler.cpp:
```diff
@@ -29,273 +32,720 @@ SliproadHandler::SliproadHandler(const IntersectionGenerator &intersection_gener
+            const auto &candidate_data = node_based_graph.GetEdgeData(candidate_road.eid);
+
+            // Name mismatch: check roads at `c` and `d` for same name
+            const auto name_mismatch = [&](const NameID road_name_id) {
```
are you considering `name_a` -> `name_b` announcement does not necessarily require `name_b -> name_a` announcement.
Checking for `EmptyNameID` is missing here and should probably be here.

MoKob commented on src/extractor/guidance/sliproad_handler.cpp:
```diff
@@ -29,273 +32,720 @@ SliproadHandler::SliproadHandler(const IntersectionGenerator &intersection_gener
             {
-                const auto &candidate_data = node_based_graph.GetEdgeData(candidate_road.eid);
-                if (target_road_names.count(candidate_data.name_id) > 0)
+                const auto skip_traffic_light_intersection = intersection_generator(
```
should probably also use `GetActualNextIntersection` from the intersection Generator

MoKob commented on src/extractor/guidance/sliproad_handler.cpp:
```diff
@@ -29,273 +32,720 @@ SliproadHandler::SliproadHandler(const IntersectionGenerator &intersection_gener
+    {
+        geojson_points.Write(std::vector<NodeID>{b, c, d});
+        geojson_lines.Write(std::vector<NodeID>{b, d});
+    }
```
remove before merge

MoKob commented on src/extractor/guidance/sliproad_handler.cpp:
```diff
@@ -29,273 +32,720 @@ SliproadHandler::SliproadHandler(const IntersectionGenerator &intersection_gener
+//  ^ at
+//     ^ road
+boost::optional<SliproadHandler::IntersectionAndNode>
+SliproadHandler::getNextIntersection(const NodeID at, const ConnectedRoad &road) const
```
can potentially be removed in favour of the intersection generator function. Feel free to adjust that part to also return node/intersection.

MoKob commented on src/extractor/guidance/sliproad_handler.cpp:
```diff
@@ -29,273 +32,720 @@ SliproadHandler::SliproadHandler(const IntersectionGenerator &intersection_gener
+    const auto forking = intersection[1].instruction.type == TurnType::Fork &&
+                         intersection[2].instruction.type == TurnType::Fork;
+
+    if (!forking)
```
am I reading this correctly? You rely on forks to have a obvious index? That sounds strange to me.

MoKob replied:
Fork should always assign both roads

MoKob commented on src/extractor/guidance/sliproad_handler.cpp:
```diff
@@ -29,273 +32,720 @@ SliproadHandler::SliproadHandler(const IntersectionGenerator &intersection_gener
+
+bool SliproadHandler::nextIntersectionIsTooFarAway(const NodeID start, const EdgeID onto) const
+{
+    BOOST_ASSERT(start != SPECIAL_NODEID);
```
many of these functions you use compute the next intersection. Given that this implies computing a new intersection, I'd prefer if we can reduce that check to a single computation. After all that is all I am trying right now to improve performance again.

MoKob commented on src/extractor/guidance/sliproad_handler.cpp:
```diff
@@ -29,273 +32,720 @@ SliproadHandler::SliproadHandler(const IntersectionGenerator &intersection_gener
+    return accumulator.too_far_away;
+}
+
+bool SliproadHandler::isThroughStreet(const EdgeID from, const Intersection &intersection) const
```
don't we have such a check somewhere else already? If not, I'd move it to the BaseClass

MoKob replied:
hm. ok. might be a good item to add a refactor ticket for to consolidate these

MoKob commented on src/extractor/guidance/sliproad_handler.cpp:
```diff
@@ -29,273 +32,720 @@ SliproadHandler::SliproadHandler(const IntersectionGenerator &intersection_gener
+
+bool SliproadHandler::roadContinues(const EdgeID current, const EdgeID next) const
+{
+    const auto current_data = node_based_graph.GetEdgeData(current);
```
references?

MoKob commented on src/extractor/guidance/sliproad_handler.cpp:
```diff
@@ -29,273 +32,720 @@ SliproadHandler::SliproadHandler(const IntersectionGenerator &intersection_gener
+}
+
+bool SliproadHandler::isValidSliproadArea(const double max_area,
+                                          const NodeID a,
```
please no single char variables :(

MoKob commented on src/extractor/guidance/sliproad_handler.cpp:
```diff
@@ -29,273 +32,720 @@ SliproadHandler::SliproadHandler(const IntersectionGenerator &intersection_gener
+}
+
+bool SliproadHandler::isValidSliproadLink(const ConnectedRoad &sliproad,
+                                          const ConnectedRoad &first,
```
first/second don't really provide an idea of what other roads you are passing in.

MoKob commented on src/extractor/guidance/sliproad_handler.cpp:
```diff
@@ -29,273 +32,720 @@ SliproadHandler::SliproadHandler(const IntersectionGenerator &intersection_gener
+    case RoadPriorityClass::LINK_ROAD:
+        factor = 0.3;
+        break;
+    case RoadPriorityClass::CONNECTIVITY:
```
You explicitly forbid Connectivity Roads above. They are never going to be sliproads. You could probably assert this here?

MoKob commented on src/extractor/guidance/sliproad_handler.cpp:
```diff
@@ -29,273 +32,720 @@ SliproadHandler::SliproadHandler(const IntersectionGenerator &intersection_gener
+        factor = 0.3;
+        break;
+    case RoadPriorityClass::CONNECTIVITY:
+        factor = 0.1;
```
are these in relation to the differences we usually assume in speed? So do all of them relate to `speed_class/speed_motorway`?

daniel-j-h commented:
After rebasing onto your changes from today @MoKob the Sliproad handler still works on Intersection but the turn handler now only knows how to find the next IntersectionView along a road which again creates incompatibilities.

Going from an IntersectionView -> Intersection happens in turn analysis, but this is where we call the Sliproad handler to begin with. And the Sliproad handler has to work on Intersections since it potentially modifies the turn type.

:(

MoKob commented on include/extractor/guidance/sliproad_handler.hpp:
```diff
@@ -42,6 +47,40 @@ class SliproadHandler : public IntersectionHandler
+    bool nextIntersectionIsTooFarAway(const NodeID start, const EdgeID onto) const;
+
+    // Through street: does a road continue with from's name at the intersection
+    bool isThroughStreet(const EdgeID from, const IntersectionView &intersection) const;
```
can you add a ticket to the guidance refactor project for removing this duplication with the intersection handler?

MoKob commented on src/extractor/guidance/motorway_handler.cpp:
```diff
@@ -490,7 +490,7 @@ Intersection MotorwayHandler::fromRamp(const EdgeID via_eid, Intersection inters
     else
     { // FALLBACK, hopefully this should never been reached
         util::Log(logDEBUG) << "Reached fallback on motorway ramp with " << intersection.size()
```
heads up: these loggings have just been removed, since we don't need to check them anymore. Only extreme cases that we don't really care about. We don't need to clutter our logs for these

MoKob commented on src/extractor/guidance/sliproad_handler.cpp:
```diff
@@ -2,15 +2,18 @@
 
-using EdgeData = osrm::util::NodeBasedDynamicGraph::EdgeData;
-using osrm::util::guidance::getTurnDirection;
 using osrm::util::guidance::angularDeviation;
```
also: moved (I believe to `#include "util/bearings.hpp"`)

MoKob commented on src/extractor/guidance/sliproad_handler.cpp:
```diff
@@ -29,273 +32,730 @@ SliproadHandler::SliproadHandler(const IntersectionGenerator &intersection_gener
-                return true;
-            };
+    // The intersection `c` has to be reachable
+    if (!main_road.entry_allowed)
```
is this not already part of `obvious`? If not, shouldn't it be?

MoKob commented on src/extractor/guidance/sliproad_handler.cpp:
```diff
@@ -29,273 +32,730 @@ SliproadHandler::SliproadHandler(const IntersectionGenerator &intersection_gener
-    const auto check_valid = [this, source_edge_data](const ConnectedRoad &road) {
-        const auto road_edge_data = node_based_graph.GetEdgeData(road.eid);
-        // Test to see if the source edge and the one we're looking at are the same road
+        auto is_roundabout = road_data.roundabout;
```
doing roundabout checks early is usually a good idea. Make sure they are up as far as possible in the processing order, since they are relatively cheap compared to walking the graph.

MoKob commented on src/extractor/guidance/sliproad_handler.cpp:
```diff
@@ -29,273 +32,730 @@ SliproadHandler::SliproadHandler(const IntersectionGenerator &intersection_gener
+        const auto &crossing_road_data = node_based_graph.GetEdgeData(crossing_road.eid);
+
+        // The crossing road at the main road's intersection must not be incoming-only
+        if (crossing_road_data.reversed)
```
I am asking myself if we could get away with checking left/right of obvious and not iterating over all roads here.

MoKob commented on src/extractor/guidance/sliproad_handler.cpp:
```diff
@@ -29,273 +32,730 @@ SliproadHandler::SliproadHandler(const IntersectionGenerator &intersection_gener
+        }(intersection_node_id);
+
+        // Constrain the Sliproad's target to sliproad, outgoing, incoming from main intersection
+        if (target_intersection.size() != 3)
```
looks like a prime candidate to check early (node_based_graph.getOutDegree(target_node) == 3)

daniel-j-h replied:
Can you please re-check the code it's using the graph walker now - which gives me back the intersection on which this check is getting performed on.

MoKob commented on src/extractor/guidance/sliproad_handler.cpp:
```diff
@@ -29,273 +32,730 @@ SliproadHandler::SliproadHandler(const IntersectionGenerator &intersection_gener
 }
 
+// Implementation details
+
```
the function calls above are only showing brief comments, could you maybe add some more specific documentation down here?

daniel-j-h replied:
The interface (fun decls in the header) should have all documentation the callsite has to know about. I think.

daniel-j-h commented:
Rebased, squashed, adapted to review remarks.

Ticketed some possible performance improvements for all intersection handlers in https://github.com/Project-OSRM/osrm-backend/issues/3416.
